dev/dram/device.c

```diff
@@ -93,75 +112,429 @@ void dram_cbt_exit(dram_dev_t *dram,
+
+    create_ck_packet_sequence(buffer, time_offset);
+
+    dram_clk = initial_dram_offset + cycles_per_packet;
+
+    create_packet_group_info(&wrdata_info, ratio, dram_clk, cycles_per_packet);
+
+    // Generate DQ data packets for setting new CA VREF value
+    // First Packet is setting DQ pins to CA VREF value
+    packet = create_ck_packet_sequence(buffer, 1);
+    if (packet == NULL)
+    {
+        return WDDR_ERROR;
+    }
+
+    fill_wrdata_packet(packet, &wrdata_info, cs, &data, cycles_per_packet, cycles_per_packet);
+
+    // Second packet is toggling DQS tDStrain after setting DQ pins
+    // Need to extend offset by 2 to account for preamble
+    dram_clk += dram->cfg->t_sh_train;
+    create_packet_group_info(&wrdata_en_info, ratio, dram_clk, 4);
+    create_packet_group_info(&wrdata_info, ratio, dram_clk, PACKET_MAX_NUM_PHASES);
+
+    if (wrdata_en_info.cycle_offset > 0)
+    {
+        wrdata_en_info.dfi_ts_start++;
+        wrdata_en_info.cycle_offset = 0;
+        wrdata_info.dfi_ts_start++;
+        wrdata_info.cycle_offset = 0;
+    }
+
+    time_offset = wrdata_en_info.dfi_ts_start - buffer->ts_last_packet;
+    do
+    {
+        packet = create_ck_packet_sequence(buffer, time_offset);
+        if (packet == NULL)
+        {
+            return WDDR_ERROR;
+        }
+
+        fill_wrdata_packet(packet, &wrdata_info, cs, &data, cycles_per_packet, PACKET_MAX_NUM_PHASES);
+        fill_wrdata_en_packet(packet, &wrdata_en_info, 4);
+
+        time_offset = 1;
+    } while (wrdata_en_info.phase_remaining);
+
+    dram_clk = (buffer->ts_last_packet + time_offset) << ratio;
+    create_packet_group_info(&wrdata_info, ratio, dram_clk, cycles_per_packet);
+    // Third Packet is setting DQ pins to CA VREF value immediately after DQS toggle
+    packet = create_ck_packet_sequence(buffer, time_offset);
+    if (packet == NULL)
+    {
+        return WDDR_ERROR;
+    }
+
+    fill_wrdata_packet(packet, &wrdata_info, cs, &data, cycles_per_packet, cycles_per_packet);
+
+    // Send CK packet tDHtrain + phy_wr_en after last packet
+    time_offset = ((dram->cfg->t_sh_train + dram->cfg->phy_wr_en) >> ratio);
+    offset_remainder = ((dram->cfg->t_sh_train + dram->cfg->phy_wr_en) & (cycles_per_packet - 1));
+
+    // Add one extra DFI cycle if tDHtrain is odd value
+    if (offset_remainder > 0)
+    {
+        time_offset++;
+    }
+    create_ck_packet_sequence(buffer, time_offset);
+
+    // Generate CA data packets to send training sequence
+    time_offset = (dram->cfg->t_vref_ca_long - dram->cfg->t_sh_train) >> ratio;
+    offset_remainder = ((dram->cfg->t_vref_ca_long - dram->cfg->t_sh_train) & (cycles_per_packet - 1));
+    // Add one extra DFI cycle if tDHtrain is odd value
+    if (offset_remainder > 0)
+    {
+        time_offset++;
+    }
+
+    packet = create_ck_packet_sequence(buffer, time_offset);
+    if (packet == NULL)
+    {
+        return WDDR_ERROR;
+    }
```

Comment:
Let's brainstorm this.

Reply:
yeah, I did see this and not a huge fan of how it is right now.  I was debating on creating a packet_group_info_t to handle this specifically, and it would only call ck_packet_sequence, but its still somewhat overkill in a way.  We should have a small discussion on how we want to handle this appropriately.  I also wanted to update the phase setting for cs to use the fill algorithm, but again, it is extra code in the sequence, but would be cleaner.

Reply:
Since the timing here isn't strict, I would say don't worry about if it's odd or not, just increase the time offset by one always.

Reply:
Fixed.